core/base/v7/inc/ROOT/TDirectoryEntry.hxx

```diff
@@ -75,7 +70,13 @@ public:
 
 template<class U>
 std::shared_ptr<U> TDirectoryEntry::CastPointer() const {
-  if (auto ptr = fType->DynamicCast(TClass::GetClass(typeid(U)), fObj.get()))
+  using cast_t = bool(const std::type_info*, const std::type_info*, void**, unsigned);
+
+  void* ptr = fObj.get();
+  void** vtbl = reinterpret_cast<void** const*>(&typeid(U))[0];
+  cast_t* cast = reinterpret_cast<cast_t**>(vtbl)[4];
```

Comment:
It probably makes sense to add the content/semantic-information of the email exchange here.  As this is clearly rely on some internal knowledge of the compiler/library implementation (very very nice find by the way!).